### DISCUSSION TITLE
NVIDIA Docker on WSL2

### DISCUSSION THREAD

**Discussion by dalekcoffee** [JeffreyCA/spleeter-web#477] [General]
Hello all!, 

The NVIDIA documentation was giving me some trouble due to deprecated commands in ubuntu. 

I just wanted to share those exploring gpu separation using WSL2 that I was able to get it working using the following guide:
https://medium.com/htc-research-engineering-blog/nvidia-docker-on-wsl2-f891dfe34abhttps://medium.com/htc-research-engineering-blog/nvidia-docker-on-wsl2-f891dfe34ab

This worked on a host running windows 11, WSL2

The only part that lost me for a bit was step 7. It actually means to install docker onto your ubuntu distro in WSL2 then follow with the commands it gives after.

Was able to run this project and an invokeAI docker container without issue!

Just set the proper expectation that heavy jobs in spleeter will use your gpu and other gpu apps will be slow or not load depending on gpu! (plex for me)

**Comment by themrfoxjr** [comment]
I am also having problems installing the cuda toolkit, drivers, etc in windows with WSL2 and Win 11. 
The link you provided with the solution is broken, can you please share what you did?

**Reply by dalekcoffee** [reply]
Honestly, even though I got it working, the container often ended up breaking for me and I gave up.

It was something along the lines of installing another ubuntu VM in WSL2, installing cuda toolkit and some other NVIDIA stuff on it, then in docker desktop enabling resource sharing or something like that. But It broke so often I am moving away from WSL2

Idk if you care, but if you plan to host some web services another issue with WSL2 is that IP addresses are not passed in headers which prevents you from using certain security features in some containers that allows you to block specific traffic or rate limit


Im moving to Hyper-V on windows 11, and a Ubuntu VM. Though, that wont work for spleeter I think, cant pass GPU in there, but Im just going to use online AI splitter websites tbh

**Reply by JeffreyCA** [reply]
I unfortunately don't have a W11 machine w/ GPU to test, but here's the fixed link: https://medium.com/htc-research-engineering-blog/nvidia-docker-on-wsl2-f891dfe34ab

**Reply by themrfoxjr** [reply]
Thank you guys for your responses. 

After some trial and error I was able to make it work with WSL2 (Ubuntu 22.04) in Win 11. I did not use Docker-desktop for windows, but instead installed Docker in the Ubuntu VM.

Here is how I made it happen:

Start by installing Nvidia Drivers (Geforce Experience)
This will include the linux drivers required. 
You should not need to install additional linux drivers inside WSL2

Open Terminal in Win 11
Install WSL2 Ubuntu 22.04

```
wsl.exe --install Ubuntu-22.04
```

After installation, install docker

```
curl https://get.docker.com | sh
```

Set up Nvidia GPG keys - Followed this video for reference (https://www.youtube.com/watch?v=CO43b6XWHNI&list=PLXtBWyNGz7u-KukGC1bjxkvGNTpzD2y0Y)

```
distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
```
```
curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
```
```
curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
```

Install Nvidia Runtime Packages

```
sudo apt-get update
```
```
sudo apt-get install -y nvidia-docker2
```

Open another WSL window

```
sudo service docker stop
```
```
sudo service docker start
```

Run an Nvidia benchmark to confirm installation

```
docker run --gpus all nvcr.io/nvidia/k8s/cuda-sample:nbody -gpu -benchmark
```
```
nvidia-smi
```

Now to install spleeter-web

These are similar to the ones found in the installation instructions.

Concerning the .env file, this means to create a new text file and call it ".env". Not a file name and extension, simply ".env" 
Here you can put your Youtube API and also any Environment Variables as shown in the Readme instructions. Mine looks like this, so I can access it from within my home network:

To get a youtube API, google "How to get a youtube API". Its free. Once you have it, copy and paste it in the .env file

```
ALLOW_ALL_HOSTS=1
NGINX_PORT=7480
YOUTUBE_API_KEY=XXXXXXxxxXXXXxxxxXX
```

Go back to the WSL terminal

```
cd /mnt/c/users/$USER$/
```
```
git clone https://github.com/JeffreyCA/spleeter-web.git
```
```
cd spleeter-web
```
Installing in WSL2, this one is modified, instead of  instruction 'docker-compose', remove the dash to make 'docker compose'
```
sudo docker compose -f docker-compose.gpu.yml -f docker-compose.prod.yml -f docker-compose.prod.selfhost.yml up
```

Install Portainer to make managing docker easier

Follow instructions here:
https://docs.portainer.io/start/install-ce/server/docker/wsl
